components/camel-kafka/src/main/java/org/apache/camel/processor/idempotent/kafka/KafkaIdempotentRepository.java

```diff
@@ -0,0 +1,342 @@
+                log.info("Cache OK");
+            } else {
+                log.warn("Timeout waiting for cache warm-up from topic {}. Proceeding anyway. "
+                        + "Duplicate records may not be detected.", topic);
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    @Override
+    public CamelContext getCamelContext() {
+        return this.camelContext;
+    }
+
+    @Override
+    protected void doStop() {
+        // stop the thread
+        topicPoller.setRunning(false);
+        try {
+            if (topicPoller.getShutdownLatch().await(30, TimeUnit.SECONDS)) {
+                log.info("Expired waiting on topicPoller to shut down");
+            }
+        } catch (InterruptedException e) {
+            log.info("Interrupted waiting on latch: {}", e.getMessage());
+        }
+        executorService.shutdown();
+
+        try {
+            consumer.close();
+        } finally {
+            producer.close();
+        }
+    }
+
+    @Override
+    public boolean add(String key) {
+        if (cache.containsKey(key)) {
+            duplicateCount.incrementAndGet();
+            return false;
+        } else {
+            broadcastAction(key, CacheAction.add);
+            return true;
+        }
+    }
+
+    private void broadcastAction(String key, CacheAction action) {
+        try {
+            log.debug("Broadcasting action:{} for key:{}", action, key);
+            producer.send(new ProducerRecord<>(topic, key, action.toString())).get(); // sync send
+        } catch (ExecutionException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    @ManagedOperation(description = "Does the store contain the given key")
+    public boolean contains(String key) {
+        log.debug("Checking cache for key:{}", key);
+        boolean containsKey = cache.containsKey(key);
+        if (containsKey) {
+            duplicateCount.incrementAndGet();
+        }
+        return containsKey;
+    }
+
+    @Override
+    @ManagedOperation(description = "Remove the key from the store")
+    public boolean remove(String key) {
+        broadcastAction(key, CacheAction.remove);
+        return true;
+    }
+
+    @Override
+    public boolean confirm(String key) {
+        return true; // no-op
+    }
+
+    @Override
+    public void clear() {
+        broadcastAction(null, CacheAction.clear);
+    }
+
+    @ManagedOperation(description = "Number of times duplicate messages have been detected")
+    public long getDuplicateCount() {
+        return duplicateCount.get();
+    }
+
+    @ManagedOperation(description = "Number of times duplicate messages have been detected")
+    public boolean isPollerRunning() {
+        return topicPoller.getRunning();
+    }
+
+    private class TopicPoller implements Runnable {
+
+        private static final int POLL_DURATION_MS = 10;
```

Comment:
Yeah, it's a trade-off against local cache liveness. If the poll duration is too long, it messes with the idempotentConsumer eager/non-eager semantics. 

I'll make it that the local cache is updated immediately, and change to 100ms so that any peer processes will be slightly behind, but not hammering the network. This seems like a reasonable poll frequency. 

What's the most Camel way to make this configurable? The repo already has quite a lot of constructor arguments. If it was just Java, I'd add a builder for the repository class, but with XML it feels like I'm straying into adding loads of setters.